crates/ty_server/src/server/api/requests/workspace_diagnostic.rs

```diff
@@ -75,20 +90,58 @@ impl BackgroundRequestHandler for WorkspaceDiagnosticRequestHandler {
                     })
                     .collect::<Vec<_>>();
 
+                // Check if this file's diagnostics have changed since the previous request
+                if let Some(previous_result_id) = previous_results.remove(&url) {
+                    if previous_result_id == result_id {
+                        // Diagnostics haven't changed, return unchanged report
+                        items.push(WorkspaceDocumentDiagnosticReport::Unchanged(
+                            WorkspaceUnchangedDocumentDiagnosticReport {
+                                uri: url,
+                                version,
+                                unchanged_document_diagnostic_report:
+                                    UnchangedDocumentDiagnosticReport { result_id },
+                            },
+                        ));
+                        continue;
+                    }
+                }
+
+                // Diagnostics have changed or this is the first request, return full report
                 items.push(WorkspaceDocumentDiagnosticReport::Full(
                     WorkspaceFullDocumentDiagnosticReport {
                         uri: url,
                         version,
                         full_document_diagnostic_report: FullDocumentDiagnosticReport {
-                            // TODO: We don't implement result ID caching yet
-                            result_id: None,
+                            result_id: Some(result_id),
                             items: lsp_diagnostics,
                         },
                     },
                 ));
             }
         }
 
+        // Handle files that had diagnostics in previous request but no longer have any
+        // Any remaining entries in previous_results are files that were fixed
+        for (previous_url, _previous_result_id) in previous_results {
```

Comment:
nit: we could probably consume `previous_results` (via `.into_iter()`) to avoid the clone on `previous_url`

Reply:
It already uses `into_iter` implicitly. The reason we need to clone the `url` is because `key` requires an owned `Url` and we also need the URL for the `FullDiagnosticReportItem`. It might be possible to avoid the URL clone by getting the `url` from the `DocumentRef` (adding an `into_url` method or similar) but it complicates the code a fair amount which I don't think is justified